staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go

```diff
@@ -0,0 +1,193 @@
+// newTLSContent determines the next set of content for overriding the baseTLSConfig.
+func (c *DynamicServingCertificateController) newTLSContent() (*dynamicCertificateContent, error) {
+	newContent := &dynamicCertificateContent{}
+
+	currClientCABundle := c.clientCA.CurrentCABundleContent()
+	// don't remove all content.  The value was configured at one time, so continue using that.
+	// Errors reading content can be reported by lower level controllers.
+	if len(currClientCABundle) == 0 {
+		return nil, fmt.Errorf("not loading an empty client ca bundle from %q", c.clientCA.Name())
+	}
+	newContent.clientCA = caBundleContent{caBundle: currClientCABundle}
+
+	return newContent, nil
+}
+
+// syncCerts gets newTLSContent, if it has changed from the existing, the content is parsed and stored for usage in
+// GetConfigForClient.
+func (c *DynamicServingCertificateController) syncCerts() error {
+	newContent, err := c.newTLSContent()
+	if err != nil {
+		return err
+	}
+	// if the content is the same as what we currently have, we can simply skip it.  This works because we are single
+	// threaded.  If you ever make this multi-threaded, add a lock.
+	if newContent.Equals(c.currentlyServedContent) {
+		return nil
+	}
+
+	// parse new content to add to TLSConfig
+	newClientCAPool := x509.NewCertPool()
+	if len(newContent.clientCA.caBundle) > 0 {
+		newClientCAs, err := cert.ParseCertsPEM(newContent.clientCA.caBundle)
+		if err != nil {
+			return fmt.Errorf("unable to load client CA file: %v", err)
+		}
+		for i, cert := range newClientCAs {
+			klog.V(2).Infof("loaded client CA [%d/%q]: %s", i, c.clientCA.Name(), GetHumanCertDetail(cert))
+			if c.eventRecorder != nil {
+				c.eventRecorder.Eventf(nil, nil, v1.EventTypeWarning, "TLSConfigChanged", "CACertificateReload", "loaded client CA [%d/%q]: %s", i, c.clientCA.Name(), GetHumanCertDetail(cert))
+			}
+
+			newClientCAPool.AddCert(cert)
+		}
+	}
+
+	// make a copy and override the dynamic pieces which have changed.
+	newTLSConfigCopy := c.baseTLSConfig.Clone()
+	newTLSConfigCopy.ClientCAs = newClientCAPool
+
+	// store new values of content for serving.
+	c.currentServingTLSConfig.Store(newTLSConfigCopy)
+	c.currentlyServedContent = newContent // this is single threaded, so we have no locking issue
+
+	return nil
+}
+
+// Run starts the kube-apiserver and blocks until stopCh is closed.
+func (c *DynamicServingCertificateController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting DynamicServingCertificateLoader")
+	defer klog.Infof("Shutting down DynamicServingCertificateLoader")
+
+	// doesn't matter what workers say, only start one.
+	go wait.Until(c.runWorker, time.Second, stopCh)
+
+	// start timer that rechecks every minute, just in case.  this also serves to prime the controller quickly.
+	_ = wait.PollImmediateUntil(1*time.Minute, func() (bool, error) {
+		c.Enqueue()
+		return false, nil
+	}, stopCh)
+
+	<-stopCh
```

Comment:
It looks like stopCh is guaranteed to be closed here.

Reply:
everything above runs in go routines. We have to wait here.